cmd/kubelet/app/server.go

```diff
@@ -412,13 +430,218 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
+		return fmt.Errorf("unable to load kubeconfig: %v", err)
+	}
+
+	authInfo, err := getCurrentContextAuthInfo(kcfg)
+	if err != nil {
+		return fmt.Errorf("unable to get auth info from kubeconfig: %v", err)
+	}
+
+	if containsSufficientTLSInfo(authInfo) {
+		return nil
+	}
+
+	authInfo.ClientCertificate, authInfo.ClientKey, err = getCertFromAPIServer(s, authInfo.ClientCertificate, authInfo.ClientKey)
+	if err != nil {
+		return fmt.Errorf("unable to get cert from API server: %v", err)
+	}
+
+	// Marshal and write the kubeconfig to disk.
+	data, err := json.Marshal(kcfg)
+	if err != nil {
+		return fmt.Errorf("unable to marshal the kubeconfig: %v", err)
+	}
+
+	if err := ioutil.WriteFile(s.KubeConfig.Value(), data, 0644); err != nil {
+		return fmt.Errorf("unable to write the kubeconfig file at %q: %v", s.KubeConfig.Value(), err)
+	}
+
+	return nil
+}
+
+// getCurrentContextAuthInfo returns the AuthInfo object that's referenced
+// by the current context.
+// If current context or auth info name is empty, then it will return an error.
+// If AuthInfo is empty, a new auth info object will be created.
+func getCurrentContextAuthInfo(config *clientcmdapi.Config) (*clientcmdapi.AuthInfo, error) {
+	ctx, ok := config.Contexts[config.CurrentContext]
+	if !ok {
+		return nil, fmt.Errorf("unable to find current context %q", config.CurrentContext)
+	}
+
+	if ctx.AuthInfo == "" {
+		return nil, fmt.Errorf("unable to find the name of the authInfo in current context %q", config.CurrentContext)
+	}
+
+	if _, ok := config.AuthInfos[ctx.AuthInfo]; !ok {
+		config.AuthInfos[ctx.AuthInfo] = clientcmdapi.NewAuthInfo()
+	}
+	return config.AuthInfos[ctx.AuthInfo], nil
+}
+
+// TODO(yifan): More detailed check on the cert / key content.
+// CheckTLSInfo returns true if the authInfo contains client certificate data for client key data.
+// Or if the client certificate or key file exists.
+func containsSufficientTLSInfo(authInfo *clientcmdapi.AuthInfo) bool {
+	// We use '||' so that we won't override existing data in case of wrong setup.
+	if len(authInfo.ClientCertificateData) > 0 || len(authInfo.ClientKeyData) > 0 {
+		return true
+	}
+
+	if crypto.FoundCertOrKey(authInfo.ClientCertificate, authInfo.ClientKey) {
+		return true
+	}
+
+	return false
+}
+
+// getCertFromAPIServer will:
+// (1) Create a restful client for doing the certificate signing request.
+// (2) Generate key pair and certificate signing request.
+//     The private key is stored to the given path.
+// (3) Send request to API server and watch for the issued certificate.
+// (4) Once (3) succeeds, dump the certificate to the given path.
+// On failure, the key and the certificate file will be cleaned up.
+// If the certfile or keyfile is empty, then it will use the default path, respectively:
+// /var/run/kubernetes/kubelet-client.srt, /var/run/kubernetes/kubelet-client.key.
+func getCertFromAPIServer(s *options.KubeletServer, certfile, keyfile string) (certPath, keyPath string, err error) {
```

Comment:
a function with a signature like this could be used by bootstrap flows and by client cert renew flows:

```
requestClientCert(client CertificateSigningRequestInterface, nodeName string) (certData []byte, keyData []byte, err error)
```

client:
* making this an interface makes it a lot easier to test
* on bootstrap, the client would be built from `--bootstrap-kubeconfig` (or whatever ends up feeding in credentials)
* on renewal, the client would be built from `--kubeconfig`

nodeName:
* I think this is all that is needed to build the subject for the client cert request

returning data instead of paths decouples the local disk writing decision (it could even be held in-memory and never persisted if the caller wanted the cert to only last for the process lifetime of the kubelet). it also  lets us test the parts of the flow individually